tests/functional/scripts/pyi_pkg_resources_provider.py

```diff
@@ -0,0 +1,514 @@
+
+# A non-existant file-directory - should return False
+assert resource_isdir(pkgname, 'subpkg1/non-existant') is False
+
+# A source script file in package - should return False
+# NOTE: PyFrozenProvider returns False because the file does not
+# exist.
+assert resource_isdir(pkgname, '__init__.py') is False
+
+# Parent of package's top-level directory
+#  * DefaultProvider returns True
+#  * ZipProvider returns False
+#  > PyiFrozenProvider disallows jumping to parent, and returns False
+# NOTE: using .. in path is deprecated and will raise exception in
+# future pkg_resources release
+ret = resource_isdir(pkgname, '..')
+assert (is_default and ret is True) or \
+       (is_zip and ret is False) or \
+       (is_frozen and ret is False)
+
+# Parent of subpacakge's directory
+#  * DefaultProvider returns True
+#  * ZipProvider returns False
+#  > PyiFrozenProvider disallows jumping to parent, and returns False
+# NOTE: using .. in path is deprecated and will raise exception in
+# future pkg_resources release
+ret = resource_isdir(pkgname + '.subpkg1', '..')
+assert (is_default and ret is True) or \
+       (is_zip and ret is False) or \
+       (is_frozen and ret is False)
+
+
+# Submodule in main package
+ret = resource_isdir(pkgname + '.a', '.')
+assert (is_default and ret is True) or \
+       (is_zip and ret is False) or \
+       (is_frozen and ret is True)
+
+ret = resource_isdir(pkgname + '.a', '')
+assert ret is True
+
+# Submodule in subpackage
+ret = resource_isdir(pkgname + '.subpkg1.c', '.')
+assert (is_default and ret is True) or \
+       (is_zip and ret is False) or \
+       (is_frozen and ret is True)
+
+ret = resource_isdir(pkgname + '.subpkg1.c', '')
+assert ret is True
+
+
+########################################################################
+#               Validate behavior of resource_listdir()                #
+########################################################################
+# List package's top-level directory
+#  * DefaultProvider lists the directory
+#  * ZipProvider returns empty list
+#  > PyiFrozenProvider lists the directory, but does not provide source
+#    .py files
+expected = {'__init__.py', 'a.py', 'b.py', 'subpkg1', 'subpkg2', 'subpkg3'}
+
+if is_frozen:
+    expected = {x for x in expected if not x.endswith('.py')}
+
+content = resource_listdir(pkgname, '.')
+content = set(content)
+
+if '__pycache__' in content:
+    content.remove('__pycache__')  # ignore __pycache__
+
+assert (is_default and content == expected) or \
+       (is_zip and content == set()) or \
+       (is_frozen and content == expected)
+
+# List package's top-level directory, with empty path
+#  > PyiFrozenProvider lists the directory, but does not provide source
+#    .py files
+expected = {'__init__.py', 'a.py', 'b.py', 'subpkg1', 'subpkg2', 'subpkg3'}
+
+if is_frozen:
+    expected = {x for x in expected if not x.endswith('.py')}
+
+content = resource_listdir(pkgname, '')
+content = set(content)
```

Comment:
Good point. I think I'll need to revise the use of `set()` anyway, because it may be hiding the potential duplicated entries issue that you noted in another comment...